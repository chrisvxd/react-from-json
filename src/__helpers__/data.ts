import { Bun, Burger, Patty, PattyIngredient } from "./burger";
import { Components, Mapping } from "./interfaces";

export const recursiveEntry = {
  type: "Burger",
  props: {
    chain: "Wahlburger",
    bun: {
      type: "Bun",
      props: {
        variant: "sesame"
      }
    },
    cheese: true,
    children: [
      {
        type: "Patty",
        props: {
          size: "large",
          ingredient: {
            type: "PattyIngredient",
            props: {
              variant: "impossible"
            }
          }
        }
      },
      {
        type: "Patty",
        props: {
          size: "large",
          ingredient: {
            type: "PattyIngredient",
            props: {
              variant: "beef"
            }
          }
        }
      }
    ]
  }
};

export const entryWithDifferentShape = {
  _type: "Burger",
  chain: "Wahlburger",
  bun: {
    _type: "Bun",
    variant: "sesame"
  },
  cheese: true,
  children: [
    {
      _type: "Patty",
      size: "large",
      ingredient: {
        _type: "PattyIngredient",
        variant: "impossible"
      }
    },
    {
      _type: "Patty",
      size: "large",
      ingredient: {
        _type: "PattyIngredient",
        variant: "beef"
      }
    }
  ]
};

export const flatEntry = {
  type: "Burger",
  props: {
    chain: "Wahlburger",
    bun: {
      type: "ComponentLookup",
      props: {
        componentType: "Bun",
        componentIndex: 0
      }
    },
    cheese: true,
    children: [
      {
        type: "ComponentLookup",
        props: {
          componentType: "Patty",
          componentIndex: 0
        }
      },
      {
        type: "ComponentLookup",
        props: {
          componentType: "Patty",
          componentIndex: 1
        }
      }
    ]
  }
};

export const flatComponents: Components = {
  Bun: [
    {
      type: "Bun",
      props: {
        variant: "sesame"
      }
    }
  ],
  Patty: [
    {
      type: "Patty",
      props: {
        size: "large",
        ingredient: {
          type: "ComponentLookup",
          props: {
            componentType: "PattyIngredient",
            componentIndex: 0
          }
        }
      }
    },
    {
      type: "Patty",
      props: {
        size: "large",
        ingredient: {
          type: "ComponentLookup",
          props: {
            componentType: "PattyIngredient",
            componentIndex: 1
          }
        }
      }
    }
  ],
  PattyIngredient: [
    {
      type: "PattyIngredient",
      props: {
        variant: "impossible"
      }
    },
    {
      type: "PattyIngredient",
      props: {
        variant: "beef"
      }
    }
  ]
};

export const mapping: Mapping = {
  Bun,
  Burger,
  Patty,
  PattyIngredient
};
