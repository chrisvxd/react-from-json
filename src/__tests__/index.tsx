import * as React from "react";
import ReactFromJSON from "../";
import * as renderer from "react-test-renderer";

import { Mapping, Components } from "../__helpers__/interfaces";
import {
  mapping,
  recursiveEntry,
  entryWithDifferentShape,
  flatEntry,
  flatComponents
} from "../__helpers__/data";

class BurgerReactFromJSON extends ReactFromJSON<Mapping, Components> {
  render(): JSX.Element {
    return super.render();
  }
}

describe("ReactFromJSON", () => {
  it("to render a recursive entry", () => {
    const tree = renderer
      .create(<BurgerReactFromJSON entry={recursiveEntry} mapping={mapping} />)
      .toJSON();

    expect(tree).toMatchSnapshot();
  });

  it("to render a recursive entry with a non-standard shape", () => {
    const tree = renderer
      .create(
        <BurgerReactFromJSON
          entry={entryWithDifferentShape}
          mapping={mapping}
          mapProp={prop => {
            if (prop._type) {
              const { _type, ...props } = prop;

              return {
                type: _type,
                props
              };
            }

            return prop;
          }}
        />
      )
      .toJSON();

    expect(tree).toMatchSnapshot();
  });

  it("to render a flat entry with a components attribute", () => {
    const tree = renderer
      .create(
        <BurgerReactFromJSON
          entry={flatEntry}
          mapping={mapping}
          components={flatComponents}
        />
      )
      .toJSON();

    expect(tree).toMatchSnapshot();
  });
});
