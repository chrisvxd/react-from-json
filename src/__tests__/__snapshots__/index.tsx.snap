// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ReactFromJSON to render a flat entry with a components attribute 1`] = `
<div>
  <div />
  <div>
    <div>
      <div />
      sesame bun
    </div>
  </div>
  <div>
    cheese
  </div>
  <div>
    <div
      id="0"
    >
      <div>
        0
      </div>
      large patty
      <div>
        impossible
      </div>
    </div>
    <div
      id="1"
    >
      <div>
        Order: 1
      </div>
      large patty
      <div>
        beef
      </div>
    </div>
  </div>
  <div>
    <div>
      <div />
      sesame bun
    </div>
  </div>
</div>
`;

exports[`ReactFromJSON to render a recursive entry 1`] = `
<div>
  <div />
  <div>
    <div>
      <div />
      sesame bun
    </div>
  </div>
  <div>
    cheese
  </div>
  <div>
    <div
      id="undefined"
    >
      <div>
        0
      </div>
      large patty
      <div>
        impossible
      </div>
    </div>
    <div
      id="undefined"
    >
      <div>
        Order: 1
      </div>
      large patty
      <div>
        beef
      </div>
    </div>
  </div>
  <div>
    <div>
      <div />
      sesame bun
    </div>
  </div>
</div>
`;

exports[`ReactFromJSON to render a recursive entry with a non-standard shape 1`] = `
<div>
  <div />
  <div>
    <div>
      <div />
      sesame bun
    </div>
  </div>
  <div>
    cheese
  </div>
  <div>
    <div
      id="undefined"
    >
      <div>
        0
      </div>
      large patty
      <div>
        impossible
      </div>
    </div>
    <div
      id="undefined"
    >
      <div>
        Order: 1
      </div>
      large patty
      <div>
        beef
      </div>
    </div>
  </div>
  <div>
    <div>
      <div />
      sesame bun
    </div>
  </div>
</div>
`;
